### PR TITLE
Generate Swift source for CoreML models

### PR DESCRIPTION
**Pull request by a2**
Implementation of #1418

> I noticed that the `coremlc generate` subcommand now — at least as of version 1127.100.30, which comes bundled with Xcode 13.3 — supports generating Swift source files with the public access modifiers (`--public-access`):
> 
> <details>
> <summary>Output from <code>xcrun coremlc generate --help</code></summary>
> 
> ```
> $ xcrun coremlc generate --help
> usage: coremlc <command> <inputdocument> <outputpath> [options ...]
> 
> Examples:
> 
> [...]
> 
> Commands:
> 
> [...]
> 
>    generate <model.{mlmodel,mlpackage}> </desired/path/to/output/directory>
>       Generate source code interface for supplied CoreML .mlmodel or .mlpackage with the following options:
>       --outputpath </desired/path/to/output/directory> : where to place the generated interface files.
>       --language <Swift|Objective-C> : interface language.
>       --swift-version x.y : generated interface requires Swift version x.y (swift only).
>       --public-access : classes and members get public access specifier (swift only).
>       --globalmodule : generate objc compatible model classes (swift only).
>       --sdkroot : filename base for System/Library/CoreServices/SystemVersion.plist (internal, swift only). Requires --platform.
>       --platform <macos|ios|watchos|tvos|maccatalyst> : check mlmodel version compatability for this platform.
>       --deployment-target xx.yy.zz : check mlmodel version compatability for this version.
>       --container <bundle-resources|swift-package> : code generation variants (internal).
>       --no-documentation : suppress inline documentation.
>       --dry-run : don't write generated code to output directory.
> 
> [...]
> ```
> </details>
> 
> As for implementing this, and because naming is one of the two most difficult problems in programming (along with cache invalidation and off-by-one errors), I see that there is are `objc_intent_library` and `swift_intent_library` functions (Macros? I'm new to Bazel) in [_apple/resources.bzl_](https://github.com/bazelbuild/rules_apple/blob/a93ad60b28877b0cfa16abafc6bdd3fce0e66c76/apple/resources.bzl).

### CODE REVIEW

**Comment by a2** [comment]
I ended up reverting to separate implementations of `generate_mlmodel_sources` and `generate_swift_mlmodel_sources` due to the way `ctx.outputs.source` ([Obj-C](https://github.com/a2/rules_apple/blob/3e83f3455f733520dc9853a4d61dcf56bbc11fba/apple/internal/resource_rules/apple_core_ml_library.bzl#L131-L133), [Swift](https://github.com/a2/rules_apple/blob/3e83f3455f733520dc9853a4d61dcf56bbc11fba/apple/internal/resource_rules/apple_core_ml_library.bzl#L222-L224)) is used / expects to be used by the macro implementation. Although the tests passed locally (minus the documentation issue which I fixed in the latest commit), I noticed an issue with resolving the model files in my own project.

File: apple/internal/resource_rules/apple_core_ml_library.bzl [hunk 1]
```diff
@@ -153,8 +153,76 @@ as `#import my/package/MyModel.h`.
 
 This rule will also compile the `mlmodel` into an `mlmodelc` and propagate it
 upstream so that it is packaged as a resource inside the top level bundle.
+""",
+)
+
+def _swift_apple_core_ml_library_impl(ctx):
```
Comment (keith) [inline apple/internal/resource_rules/apple_core_ml_library.bzl:159]: Looking at the differences between this and the existing `_apple_core_ml_library_impl` it seems like there are very few changes. instead of creating a separate rule for this what do you think about having a parameter on that internal rule for whether it generates swift or not, I guess just something like `generates_swift = True`, defaulting to false for keeping the current behavior. And then doing the same in the macro, to either create a objc_library or swift_library as the final target.

I think overall that would be a nicer user interface, and you'd dodge your naming concern. Thoughts?

File: apple/internal/resource_rules/apple_core_ml_library.bzl [hunk 2]
```diff
@@ -114,7 +114,7 @@ apple_core_ml_library = rule(
             allow_single_file = ["mlmodel"],
             mandatory = True,
             doc = """
-Label to a single `.mlmodel` file from which to generate sources and compile into mlmodelc files.
+Label to a single `.mlmodel` or `.mlpackage` file from which to generate sources and compile into mlmodelc files.
```
Comment (keith) [inline apple/internal/resource_rules/apple_core_ml_library.bzl:117]: could you split out the mlpackage changes into a separate PR and add a fixture like the one that currently exists for mlmodel?

File: apple/internal/resource_rules/apple_core_ml_library.bzl (hunk 1 above)
Comment (a2) [reply apple/internal/resource_rules/apple_core_ml_library.bzl:159]: I had an implementation like this in one of my commits on this branch. I wasn't able to make it work in practice, though, due to a larger, critical detail between `outputs = { "source": "%{name}.swift" }` (in the Swift version) and `outputs = { "source": "%{name}.m" }` (in the Obj-C version).

is there a way to declare a rule that supports something like this? Otherwise I agree that having a "language" parameter, similar to the `apple_intent_library` rule, would be ideal.

Comment (keith) [reply apple/internal/resource_rules/apple_core_ml_library.bzl:159]: If I understand the use case you don't have to predeclare outputs like this (it's actually discouraged now). You can just call ctx.actions.declare_file with the name instead.

Comment (a2) [reply apple/internal/resource_rules/apple_core_ml_library.bzl:159]: Interesting! I'll take another look tomorrow then 😃

**Review by keith — APPROVED** [review]
this looks great! one small comment, and you might need to rebase since I merged your other PR?

File: apple/internal/resource_rules/apple_core_ml_library.bzl [hunk 3]
```diff
@@ -71,7 +75,7 @@ def _apple_core_ml_library_impl(ctx):
         features = ctx.features,
         objc_fragment = None,
         platform_type_string = str(ctx.fragments.apple.single_arch_platform.platform_type),
-        uses_swift = uses_swift,
+        uses_swift = is_swift,
```
[the remark this answers is not in the document]
Comment (keith) [reply apple/internal/resource_rules/apple_core_ml_library.bzl:78]: ah yea probably need to do:

```suggestion
        uses_swift = is_swift or (swift_support.uses_swift(deps) if deps else False),
```

**Comment by keith** [comment]
the test failures appear related

**Comment by keith** [comment]
Awesome, great work! Thanks!
